Automattic/wp-calypso
# Domains: Improve reducers in Domain Management section

This PR tries to improve Domain Management reducers code based on comments provided by @ockham:

> You could use lodash's `findIndex()` (instead of filtering) along with React's `$splice` updater here. Something like

> ```js
command[ domainName ] = { records: { $splice: [ [ index, 1 ] ] } };
```

> (See https://facebook.github.io/react/docs/update.html#nested-collections)

> As you would have to negate the search criterion, you could then also use lodash's "`matches`" callback shorthand,

> ```js
const index = findIndex( state[ domainName ].records, { id: record.id, name: record.name, data: record.data, type: record.type } );
```
If `record` doesn't contain any other properties apart from those four, you might even be able to pass `record` as the second parameter to `findIndex()`.

> Otherwise, I think you can combine the above with ES6's destructuring into something like

> ```js
function deleteDns( state, domainName, record ) {
	const { id, name, record, data } = record;
	const index = findIndex( state[ domainName ].records, { id, name, record, data } );
	const command = {
		[ domainName ]: { records: { $splice: [ [ index, 1 ] ] } };
	};
	return React.addons.update( state, command );
}
```

The same refactoring should be done for:
```js
function deleteTemporaryMailbox( state, domainName, mailbox ) {
...
}
```

### Testing
It should work as before. I added more unit tests to make sure this code works as expected. I also covered edge cases with tests like when user tries to delete mailbox or DNS record that was already removed.

1. Go to http://calypso.localhost:3000/domains/manage.
2. Select a Site.
3. Select a Domain.
4. Navigate to Email page.
5. Navigate to Email Forwarding page.
6. Delete one of email forwarding rows. It should get successfully deleted.
7. Go back to Domain Settings.
8. Navigate to Name Servers and DNS page.
9. Navigate to DNS Records page.
10. Remove one of DNS record rows. It should get successfully deleted.
